Disease antagonist

Cruix opened the pull request:
:cl: 
add: Added a new mini antagonist, the sentient disease.
/:cl:

Closes #35954

Added a new mini antagonist, a disease that evolves new symptoms and powers. It gains points for infecting multiple people at the same time, and it can use those points to purchase symptoms. The overmind can jump between all infected hosts.

Right now the only non-symptom abilities are one to make the host cough, and one to make the host sneeze. The cough will try to infect people nearby the host, while the sneeze has a longer range, but can only infect people in front of the host. The cough and sneeze symptoms are different, can be purchased separately, and act like they do for regular diseases.

Evolution menu:
![menu](https://user-images.githubusercontent.com/16478175/36654736-7e00522e-1ab6-11e8-9e80-3dd058efefe0.png)

Icons for the evolution menu, cough, and sneeze buttons:
![buttons](https://user-images.githubusercontent.com/16478175/36654743-8a4da63a-1ab6-11e8-9f41-59d455d3c806.png)

Overmind and hud sprites:
![hud](https://user-images.githubusercontent.com/16478175/36654751-935c6388-1ab6-11e8-9d31-844eaafa7d94.png)

Roundend report:
![report](https://user-images.githubusercontent.com/16478175/36654755-9898939e-1ab6-11e8-83b7-736bbe9925f2.png)

Cruix committed:
Virus antagonist initial commit

Cruix committed:
Updated disease code

Cruix committed:
Sentient virus improvements

Cruix committed:
Renamed /mob/living/var/viruses to diseases, and /mob/living/var/resistances to disease_resistances

Cruix committed:
Resolved conflicts

Cruix committed:
Added sentient virus event

Cruix committed:
Renamed VIRUS defines to DISEASE defines

Cruix committed:
Fixed bugs in rewritten disease code

Cruix committed:
Fixed advanced disease Copy()

Cruix committed:
Merge branch 'disease_stuff' of https://github.com/Cruix/-tg-station into virus_antag

Cruix committed:
Finalized disease antagonist

Cruix committed:
Resolved merge conflicts

Ordonis commented:
More disease functionality and use for virologist is good

Buggy123 commented:
we plague inc now

fludd12 commented:
This is awesome!

MoreRobustThanYou commented:
Add sentient disease only symptoms

Buggy123 commented:
Given how there's beneficial symptoms on that list, are viruses gonna have the choice to go for super deadly and virulent or helpful but slow spreading, etc? It'd be interesting, though I hope that that wouldn't end up with every single virus becoming beneficial every time.

fludd12 commented:
Maybe sentient viruses could spend points to 'unlock' a symptom, then have the ability to turn it on/off? Then, for a higher cost, you could unlock the ability to control who gets what symptoms. Reward people for spreading you, and be a detriment to those who are trying to cure you!

vuonojenmustaturska commented:
Oo, this looks nice. 

Idea: maybe have "upgraded" coughing/vomit symptoms where they cause smoke/foam to appear? We have a long-lasting foam subtype courtesy of goof that could be good for this.

WJohn commented:
Sprites are acceptable, though seriously why on earth do you play with stretch to fit do you enjoy suffering? Get a bigger monitor, it makes a world of difference when you're not playing at 1.5x scale.

Ratvar commented:
Maybe reward virus for hurting/killing hosts, so it's not "beneficial virus but observer" ?
Also, medics scanning people at event mark to kill them/suicide out of spite.
"Hey dude is Jack Brens an antag cough me once if yes if you are here"

ExcessiveUseOfCobblestone commented:
This is awesome.

However, If the objective is to survive, they're always going to be epic heal disease to at least 1 individual for greentext.

Razharas commented:
It should evolve up to zombie virus including the ability to take control over dead bodies of infected people

Cruix commented:
You can click on the name of a symptom or ability in the menu to get more information on it, including the stat changes:
![untitled](https://user-images.githubusercontent.com/16478175/36675077-e614c7fc-1aff-11e8-9332-81687b54ad3d.png)

People trying to meta with the virus is no different than any other ghost role, and if medics want to start scanning people for disease just to fuck over the virus, then they are doing the job they are supposed to do.

Becoming purely beneficial virus will take considerable time, since all the healing symptoms require a large number of infected to unlock.

There is also a two minute cooldown to buying or selling abilities, so you cannot just change your whole disease on a whim.

I would like to get this testmerged and fix the inevitable deluge of bugs before making any new powers or symptoms. Zombies are a definite possibility, as are individual symptoms, with the bonus of a "kill the viro" objective.

Spaceinaba commented:
\>kill the viro objective
hahaha I love it

Ordonis commented:
Testmerge when, plague when, handwashing when

Cyberboss reviewed (changes requested):
Will need further review

DaxDupont commented:
Plague Inc SS13 edition

davethwave commented:
This sounds like a great addition to the ghost event spawn

ExcessiveUseOfCobblestone commented:
@Cruix I like the kill viro obj but we've had "friendly side antags that give epic heals with a survive obj" (borers) before and they're no longer in the game, I'd hate for this feature to go the same route.

Cruix commented:
Borers were removed because they were buggy and no one wanted to fix them. In any case, healing symptoms take a lot of time to unlock, and in order to reach them you will have to get some of the more annoying symptoms first in order to spread.

epochayur commented:
this is seriously cool

XDTM commented:
This could definitely use a set of exclusive symptoms. 
Biosuit piercing would prevent people in hardsuits and biosuits from shrugging off everything.
Wandering spores that behave like spiderlings could infect targets through vents, and people would weld vents during a contagion.
Some late-stage symptom that would allow the contagion to brainwash a vulnerable target, maybe defined as someone who took enough brain damage.
Something that can act as an offensive measure against virus immune targets, like toxin emission.
Maybe a zombie symptom as an endgame thing.

CosmicScientist commented:
>we plague inc now

Does no-one remember pandemic?

I require Madagascar to be a part of this PR as its balance fits with the rest of the game.

davethwave commented:
On the top of my head I can think of perhaps a few things missing that could be good from the symptom list. Such as anything that could cause brain damage, perhaps investing points that could buff the virus marginally in one way such as stealth +1. Trauma inducing diseases, perhaps a medical phobia symptom, and maybe since this kind of a reference to plague inc see if there might be something interesting there that could be implemented.

Kromgar commented:
PAPA NURGLE GiVES YOU HIS BLESSING 

![image](https://user-images.githubusercontent.com/6778786/36695748-318b441c-1b10-11e8-95d3-370be66bc8b1.png)

ShadowDeath6 commented:
@cosmicscientist

tbf plague inc is much more mainstream because it's not a flash game, and it has the same basic format as pandemic considering it's mentioned that it took a ton of inspiration from it

I wouldn't be opposed to a space madagascar space ruin that spawned the disease if the person entered the ruin

also :+1: I've been waiting for this since 2015

Kromgar commented:
@ShadowDeath6 

Space Madagascar ruin that's a closed harbor

Doppeltongue commented:
"CMO, an assistant in tool storage is coughing!"
"SHUT. DOWN. EVERYTHING."

I'm assuming that this antag would always/sometimes replace the current Level 7 biohazard event?

davethwave commented:
deadspace!!!

Cruix committed:
Resolved conflicts

Cruix commented:
![coughing](https://user-images.githubusercontent.com/16478175/36708050-1843daf6-1b69-11e8-96db-3a8ac9daecd5.png)

MMMiracles commented:
You need a way to fight vaccines, as there really isn't a way to prevent people from just getting a vaccine as soon as on is available. You're pretty much dead once they start spreading around unless you get real lucky and a single person doesn't go for once since you're not intrusive enough, but even then you're stuck on a single person surrounded by immune people who'll just tell the last host to grab a vaccine. Having something like a way to re-adapt completely under a new strain with a fresh start would be a good thing to have when you need to avoid being completely eradicated.

There should probably be a way to tell the difference between potential infection targets and immune targets, so you don't end up blowing your cover of the potentially unknowing infected by sneezing on an already-immune target. 

Need a way to refresh the infection menu, as it doesn't update with stuff like newly available symptoms to purchase or infected individuals. Having to open/close the menu every time for an updated view is kinda annoying.

PKPenguin321 commented:
I really hope this is as awesome as I'm imagining it

AnturK commented on code/datums/action.dm:
```diff
@@ -572,6 +573,41 @@
+	if(timeleft == 0)
+		button.maptext = ""
+		UpdateButtonIcon()
+		STOP_PROCESSING(SSfastprocess, src)
```
Should probably also stop updating when it has no owner.

AnturK commented on code/datums/diseases/_disease.dm:
```diff
@@ -116,8 +135,19 @@
 
 /datum/disease/proc/Copy()
+	//note that stage is not copied over - the copy starts over at stage 1
+	var/static/list/copy_vars = list("name", "visibility_flags", "disease_flags", "spread_flags", "form", "desc", "agent", "spread_text",
```
VV lock this just in case.

Cruix replied:
You cannot access static proc vars from VV.

AnturK replied:
Oh right, its local, carry on.

AnturK commented on code/datums/diseases/advance/advance.dm:
```diff
@@ -31,6 +30,7 @@
 	var/list/symptoms = list() // The symptoms of the disease.
 	var/id = ""
 	var/processing = FALSE
+	var/mutable = TRUE //set to FALSE to prevent most in-game methods of altering the disease via virology
```
Is there any reason to have /advance subtype now ? As in the core diseases being static was the major difference from what i can remember.

Cruix replied:
The /advance subtype has datum symptoms, while regular diseases do not.

AnturK commented on code/modules/admin/admin_verbs.dm:
```diff
@@ -595,14 +595,17 @@ GLOBAL_LIST_INIT(admin_verbs_hideable, list(
 			SSblackbox.record_feedback("tally", "admin_verb", 1, "Remove Spell") //If you are copy-pasting this, ensure the 2nd parameter is unique to the new proc!
 
-/client/proc/give_disease(mob/T in GLOB.mob_list)
+/client/proc/give_disease(mob/living/T in GLOB.mob_list)
```
Could also change the mob list

AnturK commented on code/modules/antagonists/disease/disease_mob.dm:
```diff
@@ -0,0 +1,303 @@
+
+	var/datum/atom_hud/my_hud = GLOB.huds[DATA_HUD_SENTIENT_DISEASE]
+	my_hud.add_hud_to(src)
+	START_PROCESSING(SSfastprocess, src)
```
Listening to Moved signal would probably be better for this.

AnturK reviewed (changes requested):
Looks pretty good, just some minor stuff

drawsstuff commented:
Maybe a early, cheap, harmful symptom and beneficial symptom to let people know what tree and path you plan so a cough doesnt send everyone into a panic thinking you're gonna murderbone. Alternatively, trick people into thinking your beneficial, then replace your symptoms to be harmful from that point on out!

Ratvar commented:
add  fart as reverse lowrange sneeze

Cruix commented:
Everything else I can do.

I am not going to add new symptoms, vaccine readaptation, or the like until I know that the premise of the disease is working.

Nabski89 commented:
I look forward to being lynched because people think I am responsible for this.

KorPhaeron commented:
Fantastic

MoreRobustThanYou commented:
add a virologist traitor item that summons one

XDTM commented:
Don't

MoreRobustThanYou commented:
make it 18 TC, so the cost of a holopara

XDTM commented:
no, if traitors can buy it it'll get stale really fast, get burnt out, and plus it will be associated with "lynch viro" when the viro is the one you should turn to when the contagion appears

MoreRobustThanYou commented:
oh

ExcessiveUseOfCobblestone commented:
hi mrty!

MoreRobustThanYou commented:
i regularly leave input on PRs, don't act like seeing me is rare.

tortellinitony commented:
Try to make it use TGUI

davethwave commented:
Can't wait until this is test merged

Cruix committed:
Made cooldown buttons stop processing if they are removed from an owner.
Made sentient disease active sneeze and cough not available if the host is unconscious.
Made sentient disease menu refresh when adaptations are ready or hosts are added or removed.
Made sentient disease following use movement signals instead of fastprocess.

Cruix committed:
Added better icons to sentient disease abilities

ChuckTheSheep commented:
Threw together some nicer graphics: https://file.io/8TDphb . I like this idea, good luck!

Cruix commented:
Thanks a lot for making my codersprites look even more terrible than they already did.
I'm going to leave the virus_infected one the same because I am not a huge fan of the gas cloud, and because I want it to be white and colored with the color var so there can be multiple different colors for different viruses.

ChuckTheSheep commented:
Thank you, and no worries, use which you prefer. :)

Buggy123 commented:
>Does no-one remember pandemic?

It actually took me a moment to remember what the popular version of that game was, I could only remember Pandemic at first. So, yes, but Pandemic is no longer culturally significant so I have no reason to reference it.

Ordonis commented:
Seeing this combined with some of the recent updates to surgery/traits could make this extra entertaining

AnturK commented on code/modules/antagonists/disease/disease_mob.dm:
```diff
@@ -0,0 +1,318 @@
+/mob/camera/disease/say(message)
+	return
+
+/mob/camera/disease/Move(NewLoc, Dir = 0)
```
I feel this would make more sense on key focus level but that can be changed later.

davethwave commented:
didn't notice this got merged, can't wait to see it in action!
